Change Thir to lazily create constants

b-naber opened the pull request:
To allow `AbstractConst`s to work with the previous thir changes we made and those we want to make, i.e. to avoid problems due to `ValTree` and `ConstValue` conversions, we instead switch to a thir representation for constants that allows us to lazily create constants.

r? @oli-obk

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5521088557?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/73a22ab07ae9bcce2c540a488e1cced3f230932c/checks/5521088557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error[E0425]: cannot find value `tcx` in this scope
   --> compiler/rustc_mir_build/src/thir/cx/expr.rs:685:66
    |
685 |                         let const_literal = ty::Const::from_bits(tcx, offset as u128, var_ty);
    |                                                                  ^^^ help: you might have meant to use the available field: `self.tcx`
error[E0308]: mismatched types
   --> compiler/rustc_mir_build/src/thir/cx/expr.rs:685:87
    |
    |
685 |                         let const_literal = ty::Const::from_bits(tcx, offset as u128, var_ty);
    |                                                                                       ^^^^^^ expected struct `ParamEnvAnd`, found struct `rustc_middle::ty::Ty`
    |
    = note: expected struct `ParamEnvAnd<'_, rustc_middle::ty::Ty<'_>, >`
               found struct `rustc_middle::ty::Ty<'_>`
Some errors have detailed explanations: E0308, E0425.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `rustc_mir_build` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

b-naber commented on compiler/rustc_mir_build/src/build/expr/as_temp.rs:
```diff
@@ -70,7 +70,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         Some(Box::new(LocalInfo::StaticRef { def_id, is_thread_local: true }));
                 }
-                ExprKind::Literal { const_id: Some(def_id), .. } => {
+                ExprKind::NamedConst { def_id, .. } => {
```
This might be problematic. I'm not completely sure whether we need any other variants here. Also don't know what adding constants to local_info here does later in codegen.

lcnr replied:
we previously also did this for `ExprKind::ConstParam`, would have to figure out what it's used for myself :laughing: think you can just keep the current behavior and add a FIXME for now
```suggestion
                ExprKind::NamedConst { def_id, .. } | ExprKind::ConstParam { def_id, .. } => {
```

lcnr replied:
we also want this for `ExprKind::ConstParam`: please add this as a test:
```rust
#![feature(adt_const_params)]
#![allow(incomplete_features)]

#[derive(PartialEq, Eq)]
struct Yikes;

impl Yikes {
    fn mut_self(&mut self) {}
}

fn foo<const YIKES: Yikes>() {
    YIKES.mut_self()
}

fn main() {
    foo::<{ Yikes }>()
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fd11458a9e567c54fbb2fb43c15c9f56

b-naber commented on compiler/rustc_mir_build/src/thir/cx/expr.rs:
```diff
@@ -692,32 +674,47 @@ impl<'tcx> Cx<'tcx> {
+                        );
+                        match const_literal.val() {
+                            ty::ConstKind::Value(ConstValue::Scalar(Scalar::Int(int))) => {
+                                assert_eq!(int.size(), size);
```
I still think that the way we create the `ScalarInt`s must be the source for this problem. Note that this assertion is not triggered when compiling core.

oli-obk commented on compiler/rustc_mir_build/src/thir/cx/expr.rs:
```diff
@@ -692,32 +673,28 @@ impl<'tcx> Cx<'tcx> {
+                        let offset = self.thir.exprs.push(Expr {
                             temp_lifetime,
-                            ty: var_ty,
+                            ty: expr_ty,
```
This looks suspicious ^^ you're using the cast expression type here, which in your failing example is u64, but the literal is u32

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5521103462?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/9e73a91af2da3659989e8661ab08b0ece4138ead/checks/5521103462/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.70
   Compiling unwind v0.0.0 (/checkout/library/unwind)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Size { raw: 8 }`,
 right: `Size { raw: 4 }`', compiler/rustc_mir_build/src/build/expr/as_constant.rs:61:21
   0:     0x7f4b38c48b5c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h641e0a65e2615a08
   1:     0x7f4b38cb563e - core::fmt::write::h39191aa5431a5380
   1:     0x7f4b38cb563e - core::fmt::write::h39191aa5431a5380
   2:     0x7f4b38c38a81 - std::io::Write::write_fmt::hb1861dc9906df921
   3:     0x7f4b38c4898b - std::sys_common::backtrace::print::h0ae42f20033c9262
   4:     0x7f4b38c4d1f4 - std::panicking::default_hook::{{closure}}::hd2976bf86056b49a
   5:     0x7f4b38c4cdca - std::panicking::default_hook::hd02e8d479b982aaa
   6:     0x7f4b396a7e71 - rustc_driver[7387008bc487c86e]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7f4b38c4d906 - std::panicking::rust_panic_with_hook::hbbba1d0bbca4f6bf
   8:     0x7f4b38c4d717 - std::panicking::begin_panic_handler::{{closure}}::h2d97bb73b3b478f0
   9:     0x7f4b38c49084 - std::sys_common::backtrace::__rust_end_short_backtrace::h61f77ae4323c082a
  10:     0x7f4b38c4d3e9 - rust_begin_unwind
  11:     0x7f4b38c04a93 - core::panicking::panic_fmt::h510f640c0e57f953
  12:     0x7f4b38cb1fe8 - core::panicking::assert_failed_inner::hd0148e7e424bf90b
  13:     0x7f4b3945c76b - core[8382fcd636289ab6]::panicking::assert_failed::<rustc_target[a9fcc2fb4eaff4cb]::abi::Size, rustc_target[a9fcc2fb4eaff4cb]::abi::Size>
  14:     0x7f4b3a44e527 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_constant
  15:     0x7f4b3a44e77c - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  16:     0x7f4b3a453afb - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_rvalue
  17:     0x7f4b3a457dd2 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  18:     0x7f4b3a45748c - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_temp_inner
  19:     0x7f4b3a44e98b - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  20:     0x7f4b3a44eba5 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  21:     0x7f4b3a452def - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_rvalue
  22:     0x7f4b3a457dd2 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  23:     0x7f4b3a45748c - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_temp_inner
  24:     0x7f4b3a44e98b - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  25:     0x7f4b3a44eba5 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  26:     0x7f4b3a452def - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_rvalue
  27:     0x7f4b3a457dd2 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  28:     0x7f4b3a45748c - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_temp_inner
  29:     0x7f4b3a44e98b - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  30:     0x7f4b3a44eba5 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_operand
  31:     0x7f4b3a452d89 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::as_rvalue
  32:     0x7f4b3a457dd2 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  33:     0x7f4b3a46f04d - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::in_scope::<<rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest::{closure#0}::{closure#0}, ()>
  34:     0x7f4b3a45917a - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  35:     0x7f4b3a44b534 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::ast_block_stmts
  36:     0x7f4b3a46e0b3 - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::in_scope::<<rustc_mir_build[8a8f97961e7a2025]::build::Builder>::ast_block::{closure#2}::{closure#0}, ()>
  37:     0x7f4b3a459e8f - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  38:     0x7f4b3a46f04d - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::in_scope::<<rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest::{closure#0}::{closure#0}, ()>
  39:     0x7f4b3a45917a - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  40:     0x7f4b3a46f04d - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::in_scope::<<rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest::{closure#0}::{closure#0}, ()>
  41:     0x7f4b3a45917a - <rustc_mir_build[8a8f97961e7a2025]::build::Builder>::expr_into_dest
  42:     0x7f4b3a4469dd - rustc_mir_build[8a8f97961e7a2025]::build::construct_fn::<core[8382fcd636289ab6]::iter::adapters::chain::Chain<alloc[75caa6313e26ebc1]::vec::into_iter::IntoIter<rustc_mir_build[8a8f97961e7a2025]::build::ArgInfo>, core[8382fcd636289ab6]::iter::adapters::map::Map<core[8382fcd636289ab6]::iter::adapters::enumerate::Enumerate<core[8382fcd636289ab6]::slice::iter::Iter<rustc_hir[ab985a163b28919a]::hir::Param>>, rustc_mir_build[8a8f97961e7a2025]::build::mir_build::{closure#1}::{closure#0}>>>
  43:     0x7f4b3a50bc6c - <rustc_infer[165ca10e3594e600]::infer::InferCtxtBuilder>::enter::<rustc_middle[df7eeb591dfcb37e]::mir::Body, rustc_mir_build[8a8f97961e7a2025]::build::mir_build::{closure#1}>
  44:     0x7f4b3a444849 - rustc_mir_build[8a8f97961e7a2025]::build::mir_built
  45:     0x7f4b3ab9f3ed - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::try_execute_query::<rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt, rustc_query_system[1df88c0aa7f65bfa]::query::caches::DefaultCache<rustc_middle[df7eeb591dfcb37e]::ty::WithOptConstParam<rustc_span[f78664e4d3610ddc]::def_id::LocalDefId>, &rustc_data_structures[67c22d8712ae8803]::steal::Steal<rustc_middle[df7eeb591dfcb37e]::mir::Body>>>
  46:     0x7f4b3ace3d57 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::mir_built, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  47:     0x7f4b39c4e665 - rustc_mir_transform[4c14bca07fdccc7e]::check_unsafety::unsafety_check_result
  48:     0x7f4b39c4b86f - <rustc_mir_transform[4c14bca07fdccc7e]::check_unsafety::provide::{closure#0} as core[8382fcd636289ab6]::ops::function::FnOnce<(rustc_middle[df7eeb591dfcb37e]::ty::context::TyCtxt, rustc_span[f78664e4d3610ddc]::def_id::LocalDefId)>>::call_once
  49:     0x7f4b3abb0342 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::try_execute_query::<rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt, rustc_query_system[1df88c0aa7f65bfa]::query::caches::DefaultCache<rustc_span[f78664e4d3610ddc]::def_id::LocalDefId, &rustc_middle[df7eeb591dfcb37e]::mir::query::UnsafetyCheckResult>>
  50:     0x7f4b3aca7cf0 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::unsafety_check_result, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  51:     0x7f4b39c1f420 - rustc_mir_transform[4c14bca07fdccc7e]::mir_const
  52:     0x7f4b3ab9f3ed - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::try_execute_query::<rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt, rustc_query_system[1df88c0aa7f65bfa]::query::caches::DefaultCache<rustc_middle[df7eeb591dfcb37e]::ty::WithOptConstParam<rustc_span[f78664e4d3610ddc]::def_id::LocalDefId>, &rustc_data_structures[67c22d8712ae8803]::steal::Steal<rustc_middle[df7eeb591dfcb37e]::mir::Body>>>
  53:     0x7f4b3ace4113 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::mir_const, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  54:     0x7f4b39c2017c - rustc_mir_transform[4c14bca07fdccc7e]::mir_promoted
  55:     0x7f4b3ac7642a - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::mir_promoted, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  56:     0x7f4b3a760e7e - rustc_borrowck[d83276d95930d30b]::mir_borrowck
  57:     0x7f4b3a72e55f - <rustc_borrowck[d83276d95930d30b]::provide::{closure#0} as core[8382fcd636289ab6]::ops::function::FnOnce<(rustc_middle[df7eeb591dfcb37e]::ty::context::TyCtxt, rustc_span[f78664e4d3610ddc]::def_id::LocalDefId)>>::call_once
  58:     0x7f4b3abaf612 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::try_execute_query::<rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt, rustc_query_system[1df88c0aa7f65bfa]::query::caches::DefaultCache<rustc_span[f78664e4d3610ddc]::def_id::LocalDefId, &rustc_middle[df7eeb591dfcb37e]::mir::query::BorrowCheckResult>>
  59:     0x7f4b3ac756e0 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::mir_borrowck, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  60:     0x7f4b398b589e - <rustc_middle[df7eeb591dfcb37e]::hir::map::Map>::par_body_owners::<rustc_interface[7a474fd684de44d0]::passes::analysis::{closure#2}::{closure#0}>
  61:     0x7f4b39813870 - <rustc_session[d8904b8db467f383]::session::Session>::time::<(), rustc_interface[7a474fd684de44d0]::passes::analysis::{closure#2}>
  62:     0x7f4b398ace2b - rustc_interface[7a474fd684de44d0]::passes::analysis
  63:     0x7f4b3abe942b - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::try_execute_query::<rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt, rustc_query_system[1df88c0aa7f65bfa]::query::caches::DefaultCache<(), core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>>
  64:     0x7f4b3ace0279 - rustc_query_system[1df88c0aa7f65bfa]::query::plumbing::get_query::<rustc_query_impl[97ef62945917a41b]::queries::analysis, rustc_query_impl[97ef62945917a41b]::plumbing::QueryCtxt>
  65:     0x7f4b39728084 - <rustc_interface[7a474fd684de44d0]::passes::QueryContext>::enter::<rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>
  66:     0x7f4b39715ca2 - <rustc_interface[7a474fd684de44d0]::interface::Compiler>::enter::<rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}::{closure#2}, core[8382fcd636289ab6]::result::Result<core[8382fcd636289ab6]::option::Option<rustc_interface[7a474fd684de44d0]::queries::Linker>, rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>
  67:     0x7f4b397229a9 - rustc_span[f78664e4d3610ddc]::with_source_map::<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_interface[7a474fd684de44d0]::interface::create_compiler_and_run<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}>::{closure#1}>
  68:     0x7f4b397134dd - <scoped_tls[6532d0c6e5128321]::ScopedKey<rustc_span[f78664e4d3610ddc]::SessionGlobals>>::set::<rustc_interface[7a474fd684de44d0]::interface::run_compiler<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>
  69:     0x7f4b396ce639 - std[9f788053d5631f34]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[7a474fd684de44d0]::util::run_in_thread_pool_with_globals<rustc_interface[7a474fd684de44d0]::interface::run_compiler<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>
  70:     0x7f4b3972c351 - std[9f788053d5631f34]::panic::catch_unwind::<core[8382fcd636289ab6]::panic::unwind_safe::AssertUnwindSafe<<std[9f788053d5631f34]::thread::Builder>::spawn_unchecked_<rustc_interface[7a474fd684de44d0]::util::run_in_thread_pool_with_globals<rustc_interface[7a474fd684de44d0]::interface::run_compiler<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>
  71:     0x7f4b3971d63b - <<std[9f788053d5631f34]::thread::Builder>::spawn_unchecked_<rustc_interface[7a474fd684de44d0]::util::run_in_thread_pool_with_globals<rustc_interface[7a474fd684de44d0]::interface::run_compiler<core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>, rustc_driver[7387008bc487c86e]::run_compiler::{closure#1}>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>::{closure#0}, core[8382fcd636289ab6]::result::Result<(), rustc_errors[aa610a6b17ea2182]::ErrorGuaranteed>>::{closure#1} as core[8382fcd636289ab6]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  72:     0x7f4b38c5c783 - std::sys::unix::thread::Thread::new::thread_start::hd363d8910f104f91
  73:     0x7f4b32fcd609 - start_thread
  74:     0x7f4b38ac2163 - clone
  75:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.61.0-nightly (c27e60f18 2022-03-12) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debuginfo=0 -C debug-assertions=on -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z macro-backtrace -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -C embed-bitcode=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [mir_built] building MIR for `fmt::<impl at library/core/src/fmt/mod.rs:1275:1: 2097:2>::sign_plus`
#1 [unsafety_check_result] unsafety-checking `fmt::<impl at library/core/src/fmt/mod.rs:1275:1: 2097:2>::sign_plus`
#2 [mir_const] processing MIR for `fmt::<impl at library/core/src/fmt/mod.rs:1275:1: 2097:2>::sign_plus`
#3 [mir_promoted] processing `fmt::<impl at library/core/src/fmt/mod.rs:1275:1: 2097:2>::sign_plus`
#4 [mir_borrowck] borrow-checking `fmt::<impl at library/core/src/fmt/mod.rs:1275:1: 2097:2>::sign_plus`
#5 [analysis] running analysis passes on this crate
error: could not compile `core`
Build completed unsuccessfully in 0:04:17
```

</details>

oli-obk commented on compiler/rustc_mir_build/src/thir/cx/expr.rs:
```diff
@@ -692,32 +673,28 @@ impl<'tcx> Cx<'tcx> {
+                                    ExprKind::NamedConst { def_id: did, substs, user_ty: None };
+                                let lhs = self.thir.exprs.push(Expr {
+                                    temp_lifetime,
+                                    ty: expr_ty,
```
Same here

b-naber commented:
This is ready for review now. Shouldn't be reviewed commit by commit probably.

lcnr commented on compiler/rustc_middle/src/thir/visit.rs:
```diff
@@ -93,7 +93,7 @@ pub fn walk_expr<'a, 'tcx: 'a, V: Visitor<'a, 'tcx>>(visitor: &mut V, expr: &Exp
                 visitor.visit_expr(&visitor.thir()[value])
             }
         }
-        ConstBlock { value } => visitor.visit_const(value),
```
can you delete the `visit_const` method from the `Visitor` trait?

lcnr commented on compiler/rustc_mir_build/src/build/expr/as_constant.rs:
```diff
@@ -1,43 +1,153 @@
+        debug!("expr: {:#?}", expr);
+        // FIXME: Maybe we should try to evaluate here and only create an `Unevaluated`
+        // constant in case the evaluation fails. Need some evaluation function that
+        // allows normalization to fail.
```
that comment seems incorrect, `const_eval_resolve` allows normalization to fail, it just causes an ICE when used in the incorrect `param_env`.

lcnr commented on compiler/rustc_middle/src/thir.rs:
```diff
@@ -408,13 +409,26 @@ pub enum ExprKind<'tcx> {
-        const_id: Option<DefId>,
     },
+    ConstParam {
+        literal: ty::Const<'tcx>,
```
seems weird to use a `ty::Const` for const parameters here.

want to get back to this.

lcnr replied:
maybe use a https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.ParamConst.html instead here?

oli-obk commented:
> This is ready for review now. Shouldn't be reviewed commit by commit probably.

I guess squash the commits in that case?

lcnr commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -28,16 +30,16 @@ use std::iter;
     let tcx = infcx.tcx;
 
-    if tcx.features().generic_const_exprs {
+    if infcx.tcx.features().generic_const_exprs {
```
?
```suggestion
    if tcx.features().generic_const_exprs {
```

lcnr reviewed (commented):
a few nits

lcnr commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -312,29 +315,56 @@ impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
+
+                match expr.kind {
+                    thir::ExprKind::NamedConst { substs, .. } => substs.has_param_types_or_consts(),
+                    thir::ExprKind::ConstParam { .. } => true,
```
need to add `ExprKind::Repeat` here.

Pretty much all exprs which previously had a `ty::Const` and can still be generic rn

lcnr commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -312,29 +315,56 @@ impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
                 if !self.is_poly {
-                    visit::walk_expr(self, expr)
+                    match expr.kind {
+                        thir::ExprKind::Repeat { value, count } => {
```
why is that not part of `expr_is_poly`?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5566784937?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c98ec1aaeda3eb9cc4bd42770a648464179794d3/checks/5566784937/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0599]: no method named `visit_expr` found for reference `&IsThirPolymorphic<'a, 'tcx>` in the current scope
    |
    |
328 |                         self.visit_expr(&self.thir()[value]);
    |                              ^^^^^^^^^^ method not found in `&IsThirPolymorphic<'a, 'tcx>`
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
11  | use crate::rustc_middle::thir::visit::Visitor;
11  | use crate::rustc_middle::thir::visit::Visitor;
    |

error[E0599]: no method named `thir` found for reference `&IsThirPolymorphic<'a, 'tcx>` in the current scope
    |
    |
328 |                         self.visit_expr(&self.thir()[value]);
    |                                               ^^^^ field, not a method
    = help: items from traits can only be used if the trait is in scope
help: remove the arguments
    |
    |
328 -                         self.visit_expr(&self.thir()[value]);
328 +                         self.visit_expr(&self.thir[value]);
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
11  | use crate::rustc_middle::thir::visit::Visitor;
    |
    |

error[E0308]: mismatched types
   --> compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:327:64
    |
327 |                       thir::ExprKind::Repeat { value, count } => {
    |  ________________________________________________________________^
328 | |                         self.visit_expr(&self.thir()[value]);
329 | |                         self.is_poly |= count.has_param_types_or_consts();
    | |_____________________^ expected `bool`, found `()`

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
```

</details>

lcnr commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -312,29 +315,56 @@ impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
                 self.is_poly |= pat.ty.has_param_types_or_consts();
                 if !self.is_poly {
-                    visit::walk_pat(self, pat);
+                    match pat.kind.as_ref() {
```
nit: can you also move that into a method `pat_is_poly` for consistency?

lcnr reviewed (commented):
Alright, am pretty happy with this PR now.

I do dislike `ScalarLiteral`, it feels like a pretty bad name :thinking: can't think of a better name myself, so please take 1-2 min to think of one and if not we can merge it as is :laughing:

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit b3d71d30c9badfe525fb279c455a6e466ca746fb with merge 3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b3d71d30c9badfe525fb279c455a6e466ca746fb","merge_sha":"3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b"} -->

b-naber commented:
> Alright, am pretty happy with this PR now.
> 
> I do dislike `ScalarLiteral`, it feels like a pretty bad name 🤔 can't think of a better name myself, so please take 1-2 min to think of one and if not we can merge it as is 😆

I agree that the name is bad. I would favour something like `InternalLiteral` or `NonHirLiteral`, the latter makes the intend behind the variant clearer imo, but don't like the 'Non' there. Do you find any of those better than `ScalarLiteral`?

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5646509311?check_suite_focus=true)
Build commit: 3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b (`3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5646509311?check_suite_focus=true"},"merge_sha":"3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b"} -->

rust-timer commented:
Queued 3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b with parent 3ea44938e21f0de8ae7d4f6399a8a30f97867c70, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3ea44938e21f0de8ae7d4f6399a8a30f97867c70&end=3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b).

lcnr commented:
i think `NonHirLiteral` is a bit better, but also not too great :sweat_smile: :+1:

rust-timer commented:
Finished benchmarking commit (3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b): [comparison url](https://perf.rust-lang.org/compare.html?start=3ea44938e21f0de8ae7d4f6399a8a30f97867c70&end=3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b).

**Summary**: This benchmark run shows 7 relevant improvements 🎉 but 10 relevant regressions 😿 to instruction counts.
- Arithmetic mean of relevant regressions: 1.2%
- Arithmetic mean of relevant improvements: -2.1%
- Arithmetic mean of all relevant changes: -0.2%
- Largest improvement in instruction counts: -2.8% on `full` builds of `keccak check`
- Largest regression in instruction counts: 2.2% on `full` builds of `inflate check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

b-naber commented:
@lcnr Do you consider the perf results to be problematic?

lcnr commented:
I do quite dislike that inflate has such a noticeable regression.

Apparently both `lit_to_constant` and `specializes` now get called more frequently https://perf.rust-lang.org/detailed-query.html?sort_idx=-12&commit=3fe22ef73bbaf8f3d0ef77ff2b2cb8a3fc41088b&base_commit=3ea44938e21f0de8ae7d4f6399a8a30f97867c70&benchmark=inflate-check&scenario=full

can you try to not use a query for `lit_to_constant`? I would assume that the hashmap access might not be worth the perf benefit, especially as this is pretty much a noop for integer literals

lcnr commented:
We also get `150` more calls to `specializes` and I can't immediately tell where, I think these are actually responsible for the perf degradation. We apparently now look at one more trait than before? and have to build the specialization graph for it?

The way to go ahead here is to figure out which trait is responsible here, e.g. by adding logging to the `trait_def` query and its provider for extern items.

If you want, I can try to figure out what's going on there

b-naber commented:
I can try to investigate first, haven't really investigated a perf regression in the compiler before, so probably a good chance to learn that ^^.

b-naber commented:
So this is weird, I actually get the same number of `specializes` calls in this branch and on master. First I tried this on the non-rebased branch of this PR and the commit that this PR is based on -> same number of calls. Then rebased and updated to master and get the same results:

I ran :

`RUSTC_LOG=rustc_trait_selection::traits::specialize CARGO_BUILD_RUSTC=~/Documents/rust-local-fork/thir-abstract-const/build/x86_64-apple-darwin/stage1/bin/rustc cargo build 2> thir-changes-output-rebased.txt`

and 

`RUSTC_LOG=rustc_trait_selection::traits::specialize CARGO_BUILD_RUSTC=~/Documents/rust-local-fork/playground/build/x86_64-apple-darwin/stage1/bin/rustc cargo build 2> master-output-rebased.txt`

Then filtered for `specialize::specializes` occurrences and get the same log output for both?!

I had to slightly modify the 'inflate' repo since it contained deprecated code (got rid of `try!` and changed the range pattern `...` to `..=`). Don't think that this made any difference, but probably still worth mentioning.

How shall we proceed?

lcnr commented:
interesting :thinking: I would be surprised if these changes mattered, even if keeping them would probably have been preferable.

let's try it again, maybe it was spurious in some way?

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 96050421395ae1c3a75dc0a883e0d9317f918195 with merge 7827037e4b0cf03a1f99c331fa9ef74dd062fbea...
<!-- homu: {"type":"TryBuildStarted","head_sha":"96050421395ae1c3a75dc0a883e0d9317f918195","merge_sha":"7827037e4b0cf03a1f99c331fa9ef74dd062fbea"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5662037930?check_suite_focus=true)
Build commit: 7827037e4b0cf03a1f99c331fa9ef74dd062fbea (`7827037e4b0cf03a1f99c331fa9ef74dd062fbea`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5662037930?check_suite_focus=true"},"merge_sha":"7827037e4b0cf03a1f99c331fa9ef74dd062fbea"} -->

rust-timer commented:
Finished benchmarking commit (7827037e4b0cf03a1f99c331fa9ef74dd062fbea): [comparison url](https://perf.rust-lang.org/compare.html?start=cd2da4da379b060de7cb95e631e40e2227c1992b&end=7827037e4b0cf03a1f99c331fa9ef74dd062fbea).

**Summary**: This benchmark run shows 11 relevant improvements 🎉 but 5 relevant regressions 😿 to instruction counts.
- Arithmetic mean of relevant regressions: 0.4%
- Arithmetic mean of relevant improvements: -1.6%
- Arithmetic mean of all relevant changes: -1.0%
- Largest improvement in instruction counts: -2.8% on `full` builds of `keccak check`
- Largest regression in instruction counts: 0.5% on `incr-unchanged` builds of `deeply-nested-multi check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

b-naber commented:
hm this is very weird. How can it be that the number of calls of a query differ in perf runs? Is this a bug in rustc_perf? (cc @Mark-Simulacrum)

lcnr commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -953,12 +953,17 @@ rustc_queries! {
         desc { "converting literal to const" }
     }
 
+    query lit_to_constant(key: LitToConstInput<'tcx>) -> Result<mir::ConstantKind<'tcx>, LitToConstError> {
```
still think we want to remove that query, this method is pretty cheap and only called in the same module afaict.

Using it as a method directly seems better to me

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5665281631?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ccfbe5209e09b5829e4a45b56fb609aecca426aa/checks/5665281631/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
error: unused import: `expr::as_constant`
    --> compiler/rustc_mir_build/src/build/mod.rs:1081:11
     |
1081 | crate use expr::as_constant;
     |
     |
     = note: `-D unused-imports` implied by `-D warnings`
    Checking rustc_typeck v0.0.0 (/checkout/compiler/rustc_typeck)
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error: could not compile `rustc_mir_build` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

lcnr commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 19041d995de9e3b2e4f6be46b2d77c0f19c607f7 has been approved by `lcnr`

<!-- @bors r=lcnr 19041d995de9e3b2e4f6be46b2d77c0f19c607f7 -->
<!-- homu: {"type":"Approved","sha":"19041d995de9e3b2e4f6be46b2d77c0f19c607f7","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit 19041d995de9e3b2e4f6be46b2d77c0f19c607f7 with merge 8d8135f003b35c3e109d013b2bed9ee9496da615...
<!-- homu: {"type":"BuildStarted","head_sha":"19041d995de9e3b2e4f6be46b2d77c0f19c607f7","merge_sha":"8d8135f003b35c3e109d013b2bed9ee9496da615"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5678728414?check_suite_focus=true)
Approved by: lcnr
Pushing 8d8135f003b35c3e109d013b2bed9ee9496da615 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5678728414?check_suite_focus=true"},"merge_sha":"8d8135f003b35c3e109d013b2bed9ee9496da615"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5678728414?check_suite_focus=true)
Approved by: lcnr
Pushing 8d8135f003b35c3e109d013b2bed9ee9496da615 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5678728414?check_suite_focus=true"},"merge_sha":"8d8135f003b35c3e109d013b2bed9ee9496da615"} -->

rust-timer commented:
Finished benchmarking commit (8d8135f003b35c3e109d013b2bed9ee9496da615): [comparison url](https://perf.rust-lang.org/compare.html?start=d2df372bca13bb60979c909660e69f2451630e81&end=8d8135f003b35c3e109d013b2bed9ee9496da615).

**Summary**: This benchmark run shows 32 relevant improvements 🎉 but 13 relevant regressions 😿 to instruction counts.
- Arithmetic mean of relevant regressions: 1.6%
- Arithmetic mean of relevant improvements: -1.0%
- Arithmetic mean of all relevant changes: -0.3%
- Largest improvement in instruction counts: -2.8% on `full` builds of `keccak check`
- Largest regression in instruction counts: 3.5% on `full` builds of `tuple-stress check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression

pnkfelix commented:
Visiting for weekly performance triage.

The changes to the primary benchmarks are almost universally green here.

The secondary bnechmarks are more of a mixed bag, but I think we can accept a 3.5% regression on tuple-stress given the overall benefits.

@rustbot label: +perf-regression-triaged
